### ISSUE TITLE
--cpus limitation does not work correctly with some containers on Windows Docker

### ISSUE THREAD

**Issue by georgyturevich** [moby/moby#32014]
Hello there,

I saw in `docker stats` that we have containers which were consuming ~ 15-20% CPU
```
docker stats --no-stream
CONTAINER           CPU %               PRIV WORKING SET    NET I/O             BLOCK I/O
...
2c94b939486c        16.25%              23.63 GiB           3.15 GB / 80.1 MB   48.8 MB / 637 MB
8fc83caae394        14.62%              23.48 GiB           1.67 GB / 40.7 MB   47.5 MB / 353 MB
156d442d7f4e        17.24%              23.41 GiB           3.35 GB / 127 MB    47.9 MB / 682 MB
```

And all of them have CPU limitatiion "--cpus 0.3"

```
docker inspect --format "{{.HostConfig.NanoCpus}}" 2c94b939486c 8fc83caae394 156d442d7f4e
3e+08
3e+08
3e+08
```

`docker info` shows 64 CPUs so 15-120% is quite enormous value for the "--cpus 0.3"

All our containers are runing with this --cpus limitation and usually they are not consuming more then 2% of CPU. The main difference of that containers that they perform aggressive calculation with a lot of Memory consumption up to 100 GiB. May be it make sense.

What information I can also collect for farther investigation?

Thanks!

**Output of `docker version`:**

```
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 21:19:34 2017
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.1-cs1
 API version:  1.26 (minimum version 1.24)
 Go version:   go1.7.5
 Git commit:   8709b81
 Built:        Thu Feb  9 02:05:36 2017
 OS/Arch:      windows/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 292
 Running: 110
 Paused: 0
 Stopped: 182
Images: 2816
Server Version: 1.13.1-cs1
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.693.amd64fre.rs1_release.161220-1747)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 64
Total Memory: 1.906 TiB
Name: ... skiped ...
ID: ... skiped ...
Docker Root Dir: E:\docker_storage_1_13
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 963
 System Time: 2017-03-22T14:06:35.2023919Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
It is x1.x32large EC2 instance on Dedicated host.

**Comment by thaJeztah** [comment]
Perhaps I misunderstood, but 15% CPU is less than 0.3 (30% of a single CPU)? Or were you expecting 0.3 cpu to be 0.3 percent of all CPU's?

**Comment by georgyturevich** [comment]
@thaJeztah Hi Sebastian,

As I understand `docker stats` shows what amount of whole CPU is used by containers. I saw that Task Manager shown ~ 50-60% general server CPU usage  when that three containers were used 15% per container.

**Comment by georgyturevich** [comment]
@thaJeztah forgot to add

When I am setting "--cpus 0.3" I am expecting that it will use no more the 30% of single CPU.

**Comment by justincormack** [comment]
There is an issue with kernels after 4.8 which will be fixed in 17.03.1.

**Comment by cpuguy83** [comment]
This is real Windows.

**Comment by thaJeztah** [comment]
Stats shows 100% if a single cpu is used; in your case "full cpu" would be 6400%, perhaps that's the confusion

**Comment by georgyturevich** [comment]
@thaJeztah 

According to this code https://github.com/docker/docker/blob/eb131c5383db8cac633919f82abad86c99bffbe5/cli/command/container/stats_helpers.go#L190 it does not look like calculated value can be more then 100%

`cpus = (intervalsUsed / possIntervals) * 100.0)` were `possIntervals` was multiplied by the number of processors

**Comment by georgyturevich** [comment]
@justincormack Hi Justin,

Do we have somewhere any details about the issue?

May be we have some reference or issue number? In order we can track it?

**Comment by darrenstahlmsft** [comment]
@georgyturevich I just saw this now, sorry (feel free to /cc me on issues regarding runtime limits on Windows in the future).

It's a rounding issue. The requested NanoCPUs is so small that when converting from NanoCPUs to a value that the Windows container API accepts it rounds down to 0, which is interpreted to mean unlimited CPU.

As a workaround, I would suggest using `--cpu-percent=1` instead. I'll submit a PR to fix this soon.

**Comment by georgyturevich** [comment]
@darrenstahlmsft Hi Darren,

We tried to use `--cpu-percent=1` but still faced with that cpu was spiking up to 30% for that containers

Btw, do you know what Just meant when he said:
> There is an issue with kernels after 4.8 which will be fixed in 17.03.1.

**Comment by darrenstahlmsft** [comment]
@georgyturevich I'm not able to reproduce that at all. The highest spike I'm seeing is 3% when running my https://github.com/darrenstahlmsft/go-burncpu tool in the container. How long are these spikes?

I think that statement was about Linux, and was just a confusion as there is a similar bug in new Linux kernels.
